### ISSUE TITLE
A better solution for reaching the end of content (without a 404)

### ISSUE THREAD

**Issue by cfxd** [metafizzy/infinite-scroll#619]
I am using this script on a WordPress installation (I'm not using the WP plugin) to allow visitors to manually load more content on a blog by clicking a "Load more" link. Everything works as expected, but I don't like the idea that visitors have to click "Load more" one final time when there is no more content to load. Only after they click that do they discover they've reached the end of the content, which seems like poor UX.

I'm using this on my WordPress blog, which has "Next posts" and "Previous posts" links on each page (except the first and last pages) and I set `nextSelector` to be the "Previous posts" link (selector is `.post-nav .previous a`) so that each click reveals previous (older) posts. That selector is not present anywhere on the final (oldest) page of posts.

It seems that in a setup like this, there should/could be a way to check the entire page of previous posts that's being requested for a "Previous posts" link/selector. If it's not present in the newly requested page, then we know it is the final page of content and can then either remove or alter the "Load more" link without forcing visitors to click on it one additional, fruitless time.

The most flexible solution might be to add a parameter like "`endSelector`" which gets checked on each AJAX request and if the selector doesn't exist on the page that's being requested then we know this is the final AJAX request before we reach a 404 (which could trigger `errorCallback` or another such callback). This actually means smoother UX and no more 404 when we reach the end. A win win in my book.

`endSelector` could even default to the value of `nextSelector` unless another specific selector is assigned, or it could forgo checking altogether if `false` is assigned.

Anyone have feedback or care to undertake an enhancement like this?

**Comment by mdbitz** [comment]
This can be fixed bythe addition of the following short circuit to the _loadcallback


            if( $(data).find(opts.nextSelector).length === 0 ) {
                this._showdonemsg();
                opts.state.isDone = true;
                this._binding('unbind');
            }

Full function is below will have to submit a patch later today.

_loadcallback: function infscr_loadcallback(box, data, url) {
            var opts = this.options,
            callback = this.options.callback, // GLOBAL OBJECT FOR CALLBACK
            result = (opts.state.isDone) ? 'done' : (!opts.appendCallback) ? 'no-append' : 'append',
            frag;

            // if behavior is defined and this function is extended, call that instead of default
            if (!!opts.behavior && this['_loadcallback_'+opts.behavior] !== undefined) {
                this['_loadcallback_'+opts.behavior].call(this,box,data);
                return;
            }
            if( $(data).find(opts.nextSelector).length === 0 ) {
                this._showdonemsg();
                opts.state.isDone = true;
                this._binding('unbind');
            }

			switch (result) {
				case 'done':
					this._showdonemsg();
					return false;

				case 'no-append':
					if (opts.dataType === 'html') {
						data = '<div>' + data + '</div>';
						data = $(data).find(opts.itemSelector);
					}
					break;

				case 'append':
					var children = box.children();
					// if it didn't return anything
					if (children.length === 0) {
						return this._error('end');
					}

					// use a documentFragment because it works when content is going into a table or UL
					frag = document.createDocumentFragment();
					while (box[0].firstChild) {
						frag.appendChild(box[0].firstChild);
					}

					this._debug('contentSelector', $(opts.contentSelector)[0]);
					$(opts.contentSelector)[0].appendChild(frag);
					// previously, we would pass in the new DOM element as context for the callback
					// however we're now using a documentfragment, which doesn't have parents or children,
					// so the context is the contentContainer guy, and we pass in an array
					// of the elements collected as the first argument.

					data = children.get();
					break;
			}

            // loadingEnd function
            opts.loading.finished.call($(opts.contentSelector)[0],opts);

            // smooth scroll to ease in the new content
            if (opts.animate) {
                var scrollTo = $(window).scrollTop() + $(opts.loading.msg).height() + opts.extraScrollPx + 'px';
                $('html,body').animate({ scrollTop: scrollTo }, 800, function () { opts.state.isDuringAjax = false; });
            }

            if (!opts.animate) {
				// once the call is done, we can allow it again.
				opts.state.isDuringAjax = false;
			}

            callback(this, data, url);

			if (opts.prefill) {
				this._prefill();
			}
		},

**Comment by CFXd** [comment]
:+1:

**Comment by desandro** [comment]
[Infinite Scroll v3 is out!](https://infinite-scroll.com). It's an all new re-write, so take a look at [how to upgrade from v2](https://infinite-scroll.com/extras.html#upgrading-from-v2).

Infinite Scroll has added behavior to check if the last page has been loaded. See [`checkLastPage` option](https://infinite-scroll.com/options.html#checklastpage).  This prevents Infinite Scroll from attempting to load a page that is not there.

> When set to checkLastPage: true and path set to a selector string, Infinite Scroll will check if the loaded page has the path selector.
> 
> If path is set to a string with {{#}} or a function, set checkLastPage to a selector string to check for that selector.
> 
> If path is set a function, Infinite Scroll will check if that function returns a value.

I recommend upgrading to v3. Please open a new issue if you run into trouble with Infinite Scroll v3.
